tutorials/tour/_posts/2017-02-13-lower-type-bounds.md

```diff
@@ -9,50 +9,59 @@ categories: tour
 
 Here is an example where this is useful:
 
-```tut
-case class ListNode[T](h: T, t: ListNode[T]) {
-  def head: T = h
-  def tail: ListNode[T] = t
-  def prepend(elem: T): ListNode[T] =
-    ListNode(elem, this)
+```tut:fail
+trait Node[+T] {
+  def prepend(elem: T)
 }
-```
 
-The program above implements a linked list with a prepend operation. Unfortunately, this type is invariant in the type parameter of class `ListNode`; i.e. `ListNode[String]` is not a subtype of `ListNode[Any]`. With the help of [variance annotations](variances.html) we can express such a subtype semantics:
+case class ListNode[+T](h: T, t: Node[T]) extends Node[T] {
+  def prepend(elem: T) = ListNode[T](elem, this)
+  def head: T = h
+  def tail = t
+}
 
+case class Nil[+T]() extends Node[T] {
+  def prepend(elem: T) = ListNode[T](elem, this)
+}
 ```
-case class ListNode[+T](h: T, t: ListNode[T]) { ... }
-```
+This program implements a singly-linked list. `Nil` represents an empty element (i.e. an empty list). `class ListNode` is a node which contains an element of type `T` (`head`) and a reference to the rest of the list (`tail`). The `class Node` and its subtypes are covariant because we have `+T`.
 
-Unfortunately, this program does not compile, because a covariance annotation is only possible if the type variable is used only in covariant positions. Since type variable `T` appears as a parameter type of method `prepend`, this rule is broken. With the help of a *lower type bound*, though, we can implement a prepend method where `T` only appears in covariant positions.
+However, this program does _not_ compile because the parameter `elem` in `prepend` is of type `T`, which we declared *co*variant. This doesn't work because functions are *contra*variant in their parameter types and *co*variant in their result types.
 
-Here is the corresponding code:
+To fix this, we need to flip the variance of the type of the parameter `elem` in `prepend`. We do this by introducing a new type parameter `U` that has `T` as a lower type bound.
 
 ```tut
-case class ListNode[+T](h: T, t: ListNode[T]) {
+trait Node[+T] {
+  def prepend[U >: T](elem: U)
+}
+
+case class ListNode[+T](h: T, t: Node[T]) extends Node[T] {
+  def prepend[U >: T](elem: U) = ListNode[U](elem, this)
   def head: T = h
-  def tail: ListNode[T] = t
-  def prepend[U >: T](elem: U): ListNode[U] =
-    ListNode(elem, this)
+  def tail = t
+}
+
+case class Nil[+T]() extends Node[T] {
+  def prepend[U >: T](elem: U) = ListNode[U](elem, this)
 }
 ```
 
-_Note:_ the new `prepend` method has a slightly less restrictive type. It allows, for instance, to prepend an object of a supertype to an existing list. The resulting list will be a list of this supertype.
+Now we can do the following:
+```tut
+trait Mammal
+case class AfricanSwallow() extends Mammal
+case class EuropeanSwallow() extends Mammal
```

Comment:
You'd have to do e.g. `ListNode[AfricanSwallow.type]`. Yeah, OK, better as is, actually